File: py/src/mcp_mailcow/smtp_helpers.py
"""SMTP submission helpers (port 587 STARTTLS by default)."""

from __future__ import annotations

import smtplib
import ssl
from email.message import EmailMessage
from typing import Any

from .config import UserConfig


def build_message(
    *,
    sender: str,
    to: list[str],
    subject: str,
    body: str,
    body_html: str | None = None,
    cc: list[str] | None = None,
    bcc: list[str] | None = None,
    reply_to: str | None = None,
    in_reply_to: str | None = None,
    references: str | None = None,
    attachments: list[dict[str, Any]] | None = None,
) -> EmailMessage:
    msg = EmailMessage()
    msg["From"] = sender
    msg["To"] = ", ".join(to)
    msg["Subject"] = subject
    if cc:
        msg["Cc"] = ", ".join(cc)
    if bcc:
        msg["Bcc"] = ", ".join(bcc)
    if reply_to:
        msg["Reply-To"] = reply_to
    if in_reply_to:
        msg["In-Reply-To"] = in_reply_to
    if references:
        msg["References"] = references

    msg.set_content(body)
    if body_html:
        msg.add_alternative(body_html, subtype="html")

    for att in attachments or []:
        import base64

        data = base64.b64decode(att["content_b64"])
        mime = att.get("mime_type", "application/octet-stream")
        maintype, _, subtype = mime.partition("/")
        msg.add_attachment(
            data,
            maintype=maintype,
            subtype=subtype or "octet-stream",
            filename=att["filename"],
        )
    return msg


def send_via_submission(config: UserConfig, msg: EmailMessage) -> str:
    """Send msg via SMTP submission (587 STARTTLS or 465 SSL)."""
    ctx = ssl.create_default_context()
    if not config.tls_verify:
        ctx.check_hostname = False
        ctx.verify_mode = ssl.CERT_NONE

    smtp_host = config.smtp_host or config.host
    if config.smtp_port == 465:
        with smtplib.SMTP_SSL(smtp_host, config.smtp_port, context=ctx, timeout=30) as s:
            s.login(config.mail_user, config.mail_pass)
            s.send_message(msg)
    else:
        with smtplib.SMTP(smtp_host, config.smtp_port, timeout=30) as s:
            s.starttls(context=ctx)
            s.login(config.mail_user, config.mail_pass)
            s.send_message(msg)
    return msg["Message-ID"] or ""

File: node/src/user/smtp_helpers.ts
/**
 * SMTP submission helpers using nodemailer.
 */
import { hostname } from "node:os";
import nodemailer from "nodemailer";
import type { UserConfig } from "../config.js";

// Resolved once at module load. Used in the X-Sent-By-Host header
// that every outbound message carries — lets the sysadmin trace which
// machine in the fleet sent a given mail, even when the From address is
// shared across multiple agents.
const _HOSTNAME = hostname();

export interface SendOptions {
  sender: string;
  to: string[];
  subject: string;
  body: string;
  bodyHtml?: string;
  cc?: string[];
  bcc?: string[];
  replyTo?: string;
  inReplyTo?: string;
  references?: string;
  attachments?: Array<{
    filename: string;
    content_b64: string;
    mime_type: string;
  }>;
}

export interface SendResult {
  messageId: string;
  /** RFC822 raw bytes (without BCC), for IMAP APPEND to Sent. */
  raw: Buffer;
}

function _baseMail(opts: SendOptions) {
  return {
    from: opts.sender,
    to: opts.to,
    cc: opts.cc,
    replyTo: opts.replyTo,
    subject: opts.subject,
    text: opts.body,
    html: opts.bodyHtml,
    inReplyTo: opts.inReplyTo,
    references: opts.references,
    // Forensic trace: which machine of the fleet sent this. Survives
    // even when From is a shared persona mailbox used by multiple
    // agents.
    headers: { "X-Sent-By-Host": _HOSTNAME },
    attachments: opts.attachments?.map((a) => ({
      filename: a.filename,
      content: Buffer.from(a.content_b64, "base64"),
      contentType: a.mime_type,
    })),
  };
}

export async function sendViaSubmission(
  config: UserConfig,
  opts: SendOptions,
): Promise<SendResult> {
  const transporter = nodemailer.createTransport({
    host: config.smtpHost || config.host,
    port: config.smtpPort,
    secure: config.smtpPort === 465,
    auth: { user: config.mailUser, pass: config.mailPass },
    tls: { rejectUnauthorized: config.tlsVerify },
  });

  const info = await transporter.sendMail({ ..._baseMail(opts), bcc: opts.bcc });
  const messageId = (info.messageId as string) || "";

  // Compile a raw RFC822 copy without BCC for IMAP APPEND to Sent.
  // RFC 5322: BCC must not appear in the message body delivered to anyone.
  const streamTransport = nodemailer.createTransport({
    streamTransport: true,
    buffer: true,
  });
  const compiled = await streamTransport.sendMail(_baseMail(opts));
  const raw = compiled.message as Buffer;

  return { messageId, raw };
}
